### ISSUE TITLE
fast gpt2 inference

### ISSUE THREAD

**Issue by rajarsheem** [huggingface/transformers#2550]
I have a fine-tuned ```GPT2LMHeadModel``` (gpt2-medium) which I am using to run inference on large data (>60M sequences) offline. At each iteration, my input is a batch of 30 variable-length sequences which gets padded according to the max length of the batch. My current speed is around 8 secs/iter and input sequences have around 150 tokens.

I am using pytorch's ```BucketIterator``` to group sequences of similar length and minimize padding. I am using fp-16. These are increasing the throughput. The GPU I am using is Tesla V100.

Can you please suggest what all other optimizations that I can do to increase the speed?

**Comment by rajarsheem** [comment]
Have given a thought to using TorchScript but since my input sequence length changes each time, the easier tracing approach won't work.

ONNX also faced the same problem.

**Comment by schizism** [comment]
Hi rajarsheem,
Can you please share your code that doing the batch inference with variable-length sequences?

Thanks

**Comment by Damiox** [comment]
Is gpt2 model traceable?
